The technical debt has accrued so much that there are now binary
incompatibilites with Xen-4.2, which lead to triple faults in windows
guests running older PV drivers.

The "legacy" drivers make the HVMOP_set_driver_version hypercall which
has been included for donkey's years as a NOOP.  Finally now we need to
use it to enable legacy quirks mode.

The "newstyle" XenServer 6.1 drivers for Vista and newer do not make this
hypercall, but also do not use the old hacks we have in place and use
the correct upstream ABI.

diff --git a/xen/arch/x86/hvm/hvm.c b/xen/arch/x86/hvm/hvm.c
index cd37736..c997e08 100644
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -5473,6 +5473,27 @@ long do_hvm_op(unsigned long op, XEN_GUEST_HANDLE_PARAM(void) arg)
     unsigned long start_iter, mask;
     long rc = 0;
 
+    if ( is_hvm_domain(curr_d) )
+    {
+        switch ( op )
+        {
+        case HVMOP_set_driver_version:
+            /*
+             * The XenServer 6.1 "newstyle" drivers do not make this
+             * hypercall.  All older drivers do make it.
+             *
+             * Xen-4.5 introduced the HVMOP_op_mask which truncates this op
+             * into HVMOP_set_isa_irq_level.  As HVMOP_set_isa_irq_level is
+             * ineligible for continuations, we can still distinguish legacy
+             * drivers.
+             */
+            curr_d->arch.hvm_domain._win_legacy_quirks = 1;
+            gdprintk(XENLOG_INFO, "Legacy windows driver quirks enabled. "
+                     "(HVMOP_set_driver_version)\n");
+            goto legacy_win_complete;
+        }
+    }
+
     switch ( op & HVMOP_op_mask )
     {
     default:
@@ -6142,6 +6163,7 @@ long do_hvm_op(unsigned long op, XEN_GUEST_HANDLE_PARAM(void) arg)
                                            op | start_iter, arg);
     }
 
+ legacy_win_complete:
     return rc;
 }
 
diff --git a/xen/include/asm-x86/hvm/domain.h b/xen/include/asm-x86/hvm/domain.h
index 2757c7f..c9735f1 100644
--- a/xen/include/asm-x86/hvm/domain.h
+++ b/xen/include/asm-x86/hvm/domain.h
@@ -146,6 +146,9 @@ struct hvm_domain {
         struct vmx_domain vmx;
         struct svm_domain svm;
     };
+
+    /* Do we need to enable legacy windows driver quirks for this domain? */
+    bool_t                 _win_legacy_quirks;
 };
 
 #define hap_enabled(d)  ((d)->arch.hvm_domain.hap_enabled)
diff --git a/xen/include/public/hvm/hvm_op.h b/xen/include/public/hvm/hvm_op.h
index eeb0a60..955a94e 100644
--- a/xen/include/public/hvm/hvm_op.h
+++ b/xen/include/public/hvm/hvm_op.h
@@ -369,6 +369,13 @@ DEFINE_XEN_GUEST_HANDLE(xen_hvm_set_ioreq_server_state_t);
 
 #endif /* defined(__XEN__) || defined(__XEN_TOOLS__) */
 
+#define HVMOP_set_driver_version 0x103
+struct xen_hvm_set_driver_version {
+    uint32_t build;
+};
+typedef struct xen_hvm_set_driver_version xen_hvm_set_driver_version_t;
+DEFINE_XEN_GUEST_HANDLE(xen_hvm_set_driver_version_t);
+
 #endif /* __XEN_PUBLIC_HVM_HVM_OP_H__ */
 
 /*
